### PR TITLE
Create new branch from branch selection dropdown

### PR DESCRIPTION
**Pull request by lafriks**
Also fix Branches and Tags tab alignment in branch selection select box

Screenshot:
attels

### CODE REVIEW

File: models/repo.go
Comment (bkcsoft) [inline models/repo.go]: Are you sure that we need a clone to do this? Since one can create new branches in a bare repo...

File: routers/routes/routes.go
Comment (bkcsoft) [inline routers/routes/routes.go]: Why `_new` instead of `new` ? (Except that we'd need to add another reserved name)

File: models/repo.go
Comment (bkcsoft) [reply models/repo.go]: Also see https://github.com/go-gitea/gitea/blob/master/models/repo.go#L2341 for reference

File: models/repo.go
Comment (bkcsoft) [inline models/repo.go]: Please put this function in `models/repo_branch.go` instead. Let's not add more to the God File unless we have to 😄

File: routers/routes/routes.go
Comment (bkcsoft) [reply routers/routes/routes.go]: Just to clarify, I'm fairly certain that no one would create a branch called "new" anyhow 😆

**Review by bkcsoft — CHANGES_REQUESTED** [review]
and a question

File: routers/routes/routes.go
Comment (lafriks) [reply routers/routes/routes.go]: I was following same naming rules that are for editing files and I don't think we need to add new reserved name, it is completely new group under repo

File: models/repo.go
Comment (lafriks) [reply models/repo.go]: should I also move `CreateNewBranch` function that was already there to repo_branch.go? As these functions are doing very similar things I just did not want them to be in different files

File: models/repo.go
Comment (lafriks) [reply models/repo.go]: It is for creating new branch based on existing branch/tag/commit, so no bare repositories. I was basing it exactly on your pointed out code. Branch can be created directly in real repository without cloning but I just thought that there are reasons why existing branch creation from branch was made this way

File: models/repo.go
Comment (bkcsoft) [reply models/repo.go]: @lafriks Yes please 🙂

Comment (bkcsoft) [reply models/repo.go]: Maybe if possible make it the same function (renaming `oldBranch` to `baseRef` or similar)

File: models/repo.go
Comment (bkcsoft) [reply models/repo.go]: Yeah I noticed that after I commented on it. I wonder why it does that 🤔

File: models/repo.go
Comment (lafriks) [reply models/repo.go]: They have different functionality a bit and different checks and some code is called in different order, so making all one function would overcomplicate it

File: models/repo.go
Comment (lafriks) [reply models/repo.go]: That is very good question that I would really be interested to get answer, it could be just for that it is also used by file editor but could be also other reasons, hard to tell

File: templates/repo/branch_dropdown.tmpl
Comment (ethantkoenig) [inline templates/repo/branch_dropdown.tmpl]: Can just do `data-can-create-branch="{{.Repository.CanCreateBranch}}"`

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: templates/repo/branch_dropdown.tmpl
Comment (lafriks) [reply templates/repo/branch_dropdown.tmpl]: Did not work for me...

File: integrations/repo_branch_test.go
Comment (lafriks) [reply integrations/repo_branch_test.go]: Will fix, I did write this code before MakeRequest was changed and forgot to remove that line

File: routers/routes/routes.go
Comment (lafriks) [reply routers/routes/routes.go]: @ethantkoenig it would be right way but semantic-ui dropdown does not have enough events exposed to do that :(

File: templates/repo/branch_dropdown.tmpl
Comment (lafriks) [reply templates/repo/branch_dropdown.tmpl]: Fixed

File: routers/routes/routes.go
Comment (ethantkoenig) [reply routers/routes/routes.go]: @lafriks Can't we just add some Javascript/JQuery to make a POST request?

Comment (lafriks) [reply routers/routes/routes.go]: @ethantkoenig not with jquery if only doing with some js timeouts to check if semantic ui dropdown has rendered that link and than binding on click but that is quite dirty that I would want to avoid. Also it could be done using `<a onclick` with bunch of javascript but that code also seems more like hack than good solution

Comment (ethantkoenig) [reply routers/routes/routes.go]: @lafriks If we use `GET`, then there will no be CSRF check

Comment (ethantkoenig) [reply routers/routes/routes.go]: @lafriks What do you think of this: https://github.com/lafriks/gitea/pull/1?

**Comment by lafriks** [comment]
@bkcsoft, @ethantkoenig fixed with all your suggestions, please review

File: routers/repo/branch.go
Comment (bkcsoft) [inline routers/repo/branch.go]: Where is this limit taken from? Why is it not a constant? (in `go-gitea/git`) :)

File: routers/repo/branch.go
Comment (bkcsoft) [inline routers/repo/branch.go]: This should probably be `repo.branch.tag_collision`

**Review by bkcsoft — COMMENTED** [review]
1 question, 1 suggestion. Otherwise LGTM

File: routers/repo/branch.go
Comment (lafriks) [inline routers/repo/branch.go]: It was @ethantkoenig suggestion. But I'll change it to 255 to be same as GitHub's limit

Comment (lafriks) [reply routers/repo/branch.go]: @bkcsoft it is limit used also in other places in gitea code where branch name can be entered (editor for example)

Comment (bkcsoft) [reply routers/repo/branch.go]: No clue 😆 Hence why I'm asking 😉

Comment (lafriks) [reply routers/repo/branch.go]: @bkcsoft @ethantkoenig suggested that limit and later I found it also in other places in code, so that's why it is so :)

Comment (ethantkoenig) [reply routers/repo/branch.go]: I suggested it because I found it in other places :laughing:

**Comment by ethantkoenig** [comment]
As things currently are, I can't create a branch called "mast", since searing for "mast" will match the master branch (so the "Create new branch" button won't show). Maybe the button should always show, but be disabled for exact matches?

**Comment by lafriks** [comment]
You are right but with this dropdown I don't know if it is possible. I will check if I can work around somehow

**Comment by lafriks** [comment]
@ethantkoenig @bkcsoft I rewrote javascript side to allow creating branches also with names that partially match existing branches as suggested, please review

**Comment by lafriks** [comment]
As I could not solve small remaining issue that no data found would still show up sometimes together with create new branch item, I did complete JavaScript branch/tag dropdown rewrite to not use semantic ui js but VueJS. As a plus works a lot faster and in future more extendable.

**Comment by lafriks** [comment]
Tested on IE11, Edge, Chrome, Firefox (Linux & Windows)

**Comment by ethantkoenig** [comment]
If the branch `foo` exists, we need to prevent users from creating a branch `foo/bar` (there will be a 500)

**Comment by lafriks** [comment]
@ethantkoenig good catch, I will look into it

**Comment by Bwko** [comment]
LGTM

**Comment by lafriks** [comment]
@ethantkoenig @bkcsoft added better branch name validation, please review

File: routers/repo/branch.go
Comment (ethantkoenig) [inline routers/repo/branch.go]: Need to `return`

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: routers/repo/branch.go
Comment (lafriks) [reply routers/repo/branch.go]: @ethantkoenig added

**Comment by lunny** [comment]
@lafriks please rebase

File: routers/repo/branch.go
Comment (bkcsoft) [reply routers/repo/branch.go]: And I'm suggesting making that a global constant variable in `code.gitea.io/git` :trollface:

**Comment by lafriks** [comment]
@lunny I will rewrite it to more modular component as now there is code duplication in code view and issue branch/tag selection

**Review by bkcsoft — APPROVED** [review]

**Comment by lafriks** [comment]
@ethantkoenig need your approval

File: models/repo_branch.go
Comment (ethantkoenig) [inline models/repo_branch.go]: I believe this is duplicated code; this `if`/`else` can be replaced by
```
UpdateLocalCopyBranch(repoPath, localPath, repo.DefaultBranch)
```

File: models/repo_branch.go
Comment (ethantkoenig) [inline models/repo_branch.go]: Since the function should only be called when the repo's "mutex" has been acquired, should we make it private? Similarly, for `repo.UpdateLocalCopyToCommit`?

File: routers/repo/branch.go
Comment (ethantkoenig) [inline routers/repo/branch.go]: Could we move this condition to a helper function, something like `BranchNamesConflict(name1, name2)`? It could be useful in other places, and would make the code more readable IMO.

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
~It's not working when I try it locally (`eae7ad5`); the link to create a new branch never appears.~ Never mind
image

**Comment by lafriks** [comment]
@ethantkoenig from what it looks like from screenshot you are not logged in. For branch creation to show up you must be logged in and have write access to that repository

**Comment by ethantkoenig** [comment]
@lafriks Wow, not sure how I missed that :man_facepalming:. Links appearing as they should.

**Comment by webjoel** [comment]
index.css gets created by make stylesheets from less file. So shouldn't matter.
you have to call that for yourself.

can someone finish the job, I'm waiting for this feature :)

File: routers/repo/branch.go
Comment (lafriks) [reply routers/repo/branch.go]: @ethantkoenig I have no idea to where to move this functions

**Comment by lunny** [comment]
LGTM

File: models/repo_branch.go
Comment (lafriks) [reply models/repo_branch.go]: @ethantkoenig no it can not as they are different and calls methods in different order. Code from UpdateLocalCopyBranch will fail in some cases if not updated in correct order (at least for creating new branch from commit/tag)

File: modules/validation/binding.go
Comment (bkcsoft) [inline modules/validation/binding.go:52]: How about using a regex instead of doing similar checks twice? https://stackoverflow.com/a/12093994

File: routers/repo/branch.go
Comment (bkcsoft) [reply routers/repo/branch.go]: `models.(*Repository).BranchNameConflict(form.NewBranchName)` perhaps?

File: modules/validation/binding.go
Comment (lafriks) [reply modules/validation/binding.go:52]: That won't work in golang as its regex does not support all required features. I tried that exact stackoverflow topic regex first when implementing this

**Comment by lafriks** [comment]
@ethantkoenig @bkcsoft moved branch name validation to model

**Review by ethantkoenig — APPROVED** [review]

File: models/repo_branch.go
Comment (bkcsoft) [inline models/repo_branch.go]: Do this before checking the branch-name, otherwise you'll end up with race conditions

File: models/repo_branch.go
Comment (bkcsoft) [inline models/repo_branch.go]: same here

File: routers/repo/branch.go
Comment (bkcsoft) [inline routers/repo/branch.go]: so 500 error if the name is invalid?

**Review by bkcsoft — CHANGES_REQUESTED** [review]

File: models/repo_branch.go
Comment (bkcsoft) [inline models/repo_branch.go:82]: Instead of doing this, how about something like this that scales to thousands of branches?
```
if _, err := gitRepo.GetBranch(name); err != nil {
  return ErrConflict(name)
}
if strings.Contains(name, "/") {
  for index := strings.LastIndex(name, "/"); index > 0; index := strings.LastIndex(name, "/") {
    if conflict, err := gitRepo.GetBranch(name[:index]); err != nil {
      return ErrConflict(name[:index])
    }
  }
}

Comment (lafriks) [reply models/repo_branch.go:82]: It is not the same (if `name` is `feat` and there already exists branch `feat/test`) this code won't return error

File: routers/repo/branch.go
Comment (lafriks) [reply routers/repo/branch.go]: No there is L56-L73 special error type checks before this - this one is for unexpected errors

File: models/repo_branch.go
Comment (lafriks) [reply models/repo_branch.go]: @bkcsoft fixed both

File: models/repo_branch.go
Comment (ethantkoenig) [reply models/repo_branch.go:82]: One option is to just try the `git branch ...` command, and see if it succeeds or not. I assume that internally git has an efficient way to check for naming conflicts.

Comment (bkcsoft) [reply models/repo_branch.go:82]: I think that `gitRepo.CreateBranch` will throw a collision error no?

**Review by bkcsoft — APPROVED** [review]

File: models/repo_branch.go
Comment (lafriks) [reply models/repo_branch.go:82]: But then `CheckBranchName` will only provide part of validation (as I understand that that was an initial request to create it at first place to reuse it in other places later). Also currently GetBranches is called on every repo page load (in ctx `RepoAssignment`) so it's usage here anyway won't be that big deal

**Pull request merged**
